### ISSUE TITLE
Android tv pie hyperion grabber

### ISSUE THREAD

**Issue by timstipdonk** [abrenoch/hyperion-android-grabber#153]
Ever since I've updated my mibox S to android pie the hyperion screen grabber stopped working in all media apps (even the ones that did use to work, I'm aware of the DRM limitations). Do you think android pie has changed some of its api's and therefore the grabber is unable to work in certain applications?

**Comment by abrenoch** [comment]
Hey there. Nothing with the apis have changed in android pie. Perhaps there was a firmware change or something with your hardware? It's unfortunate that the app's behavior can be unpredictable across different hardware, but that is the nature of this sort of thing.

**Comment by linarense** [comment]
I have the same problem. Any solution??? Thanks for your great work.

**Comment by scotie** [comment]
Same here. After updating to Android 9 stopped working

**Comment by mchauvel** [comment]
Is there a way to provide some logs to help to debug this point ? Same problem here with Kodi and Emby

**Comment by abrenoch** [comment]
Unfortunately I don't think this is the sort of thing having logs is going to help... If the app appears to work correctly in the app drawer but nowhere else, that would mean the grabber is running, but that there must be a protected surface drawing the rest of the screen content (which seems odd). When the android system is delivering an image from a protected surface to the grabber, it comes thru as just black. Without some detailed changelog of what exactly has changed since the update (I looked around a little but came up empty), I'm not sure I can speak with any confidence on this problem.

Really stinks, wish I could offer more help right now!

There _might_ be a chance to get this running again if you can find a different ROM to run, but I'm just speculating

**Comment by linarense** [comment]
Thanks abrenoch. Could help if we root the mibox S?

**Comment by abrenoch** [comment]
It is *possible* rooting might help, if you use something like [this](https://forum.xda-developers.com/showpost.php?p=73948857). I can't guarantee anything here though. To be honest, rooting might have its own set of drawbacks as far as DRM behavior goes.

**Comment by mchauvel** [comment]
With Kodi in Video Settings, i disabled "MediaCodec (Surface)" et let enabled "MediaCodec" and grabber (and hyperion kodi pluggin) works now.
Can other mibox owner confirm this ?

**Comment by scotie** [comment]
> With Kodi in Video Settings, i disabled "MediaCodec (Surface)" et let enabled "MediaCodec" and grabber (and hyperion kodi pluggin) works now.
> Can other mibox owner confirm this ?

I couldn't find this setting in Kodi. I'm using Kodi 18.5.
P.S.
It 8s working in the Kodi menu, but not during playback

**Comment by mchauvel** [comment]
See my video settings screen (in french)
![Screenshot_20200222-122036](https://user-images.githubusercontent.com/1884317/75091639-e6322c00-556f-11ea-8383-bea784748d1c.png)

**Comment by scotie** [comment]
> See my video settings screen (in french)

Yes, this works, but it affects the playback in a bad way.

**Comment by scotie** [comment]
I've found that casting a YouTube video from my phone also works with grabber.
Pretty sure it didn't work in the previous Android version

**Comment by scotie** [comment]
Hi, any new regarding this issue?
Something I can do to help?

**Comment by coolwormgit** [comment]
same issue here, working only with smartube.... and android menu... with mxplayer only if you change to software decode.. not working with hw+ , any fix??

**Comment by scotie** [comment]
> same issue here, working only with smartube.... and android menu... with mxplayer only if you change to software decode.. not working with hw+ , any fix??

Do you mean Smart YouTube TV?
How did you get it to work? It's not working for me after Android 9 update.

As I understand Mi box s with Android 9 always has deinterlacing enabled, hence only software decoding is working

**Comment by coolwormgit** [comment]
yes smart youtube... i didnt do anything... it was the only working app...

**Comment by scotie** [comment]
Any news regarding this issue? Something to try?

**Comment by iy2koi** [comment]
Today when I was watching netflix I realized that the ambilight is on and working. Then I closed the app, opened again and it's never working again since then. So I think there has to be a service or something and if we can manage to kill or froze that, maybe we can use the framegrabber in drm protected apps.

**Comment by derzwen** [comment]
I think this grabber needs to be updated to support newer Android versions (9+) properly. 

I'm not an Android dev, but a long time java/web dev. So i've looked though your code a bit and wanted to understand what Android permissions it actually needs to record the screen and how it's done.

I think the requestPermissions method and some other smaller stuff

https://github.com/abrenoch/hyperion-android-grabber/blob/c2befd7501c72f93f5c022666266f67142fe8ba3/common/src/main/java/com/abrenoch/hyperiongrabber/common/ToggleActivity.java#L81

needs to be updated because after a bit of googling I found out, that with Android 9 the use of the MEDIA_PROJECTION stuff has changed a little. I also found a interesting stackoverflow post about this with a couple of hints.

https://stackoverflow.com/questions/61276730/media-projections-require-a-foreground-service-of-type-serviceinfo-foreground-se

@abrenoch Can you please take a look a this? As I said, i'm not an Android dev, but this sounds at least like a direction how to get this working again under Android 9+.

I'm also willing to help testing this with my 2019 Shield pro if you provide me an apk (the current grabber version doesn't work because it's using the latest shield software which is Android 9+)

**Comment by pirti001** [comment]
Today i tried Hyperion.ng + Google Chromecast 4 and it's working ! maybe because Chromecast 4 using Android 10 ? well   youtube app playback isn't smooth playback :( but leds works and Kodi app working awesome only for 1080p ! I didn't have to turn off MediaCodec (Surface) and for 4k resolution isn't smooth playback but problem is maybe in my hyperion.ng settings because in my capturing hardware i can't set 4k resolution, i can choose only 1080p, one times i had similar problem but i forgot how solved it. Recommended use Android version hyperion tv !!!

**Comment by PeterMalast** [comment]
> Today i tried Hyperion.ng + Google Chromecast 4 and it's working ! maybe because Chromecast 4 using Android 10 ? well youtube app playback isn't smooth playback :( but leds works and Kodi app working awesome only for 1080p ! I didn't have to turn off MediaCodec (Surface) and for 4k resolution isn't smooth playback but problem is maybe in my hyperion.ng settings because in my capturing hardware i can't set 4k resolution, i can choose only 1080p, one times i had similar problem but i forgot how solved it. Recommended use Android version hyperion tv !!!

**Comment by ihrapsa** [comment]
Any news on mibox s? It's weird that only the menus work even if the content is not drm protected.

**Comment by psbankar** [comment]
I have Mi TV from Xiaomi running Android Pie and screen grab works fine in it. But in mibox, even common apps like YouTube, Plex don't work

**Comment by Nosskirneh** [comment]
It worked for me until very recently. No recent software update installed. Now it just stops when I start playing any media it seems like.

**Comment by thehijacker** [comment]
Same issue here with Mi Stick 4K. Menu is working but no played video is shown. No matter with which app (DRM or not). Anyone figure this out?

**Comment by TheNewsYorkTime** [comment]
I have an AirTv mini with sling stuff disabled. Whenever I try to screen grab a video on Netflix, Disney Plus, or YouTube the lights turn off. Even casting YouTube to the TV is broken. Casting a normal video works, as does the splash screen and login page of Disney Plus. Amazon menus work. Netflix doesn't work at all. So do YouTube menus. I'm running android 9.
